mantidproject/mantid
# Update Reflectometry algorithms to take optional polarization efficiencies workspace

**Description of work.**

Makes the algorithm changes required for issue #34744. The GUI side of the work will be added in a separate PR.

`ReflectometryReductionOneAuto` now optionally takes a workspace containing the polarization analysis efficiency factors so that values can now be supplied this way instead of via the parameter file, if preferred.

`ReflectometryISISLoadAndProcess` has been updated so that the user can supply the name of either a workspace in the ADS or a file name to be loaded and passed to `ReflectometryReductionOneAuto` as the polarization efficiency workspace.

The method of polarization correction to be applied is inferred from the efficiency workspace labels, although I've left full validation of the efficiency workspace to the algorithms that it is passed to later in the process. The correction method (analysis mode/flipper configuration) is inferred from the input workspace group.

As explained on the issue, the changes are required to be available in a nightly build by 9th December. I've added unit tests but there is quite a bit of duplication in the tests for `ReflectometryReductionOneAuto`. Given the time constraint, I felt this could be improved in a follow-up PR to allow the rest of the work to progress.

**To test:**

1) Run the `ReflectometryISISLoadAndProcess` algorithm using the POLREF test data provided by Andrew (check with Rachel about this). Test running the algorithm with `PolarizationAnalysis` ticked and the test file name entered into `PolarizationEfficiencies`. There should be two IvsLam workspaces in the IvsLam output group, one with suffix `_--` and the other with suffix `_++`.
1) Load the test file name as a workspace into the ADS with a new name. Run the algorithm again using the workspace name for the `PolarizationEfficiencies` parameters to check it is picked up from the ADS.
1) Check that sensible error messages are displayed if an invalid name is provided for `PolarizationEfficiencies`.
1) Check that polarization correction is not applied if the `PolarizationAnalysis` option is unticked, regardless of whether or not a value is provided to `PolarizationEfficiencies`. This can be checked by looking at the suffixes for workspaces in the IvsLam group (see step 1).
1) Check that with `PolarizationAnalysis` ticked and `PolarizationEfficiencies` left blank, the efficiency factors are looked up from the parameter file (POLREF doesn't have the necessary information defined in the parameter file, so you should expect this to fail with error "Correction method is undefined").
1) Check that running a normal reduction and preview reduction via the interface still works as it did previously.


Refs #34744. 



*This does not require release notes* because **they should be added in the PR that includes the GUI side of the changes.**



---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
